[EOC-142] Cloning item

alemikolo opened the pull request:
1.  If cloning the last item causes that the items count is above the limit of displayed items, this cloned item is being unmounted. Since there is still async action in progress we get warning in the browser. To prevent it I reject pending promises in componentWillUnmount. This solution is based on (https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html) and (https://medium.com/trabe/avoid-updates-on-unmounted-react-components-2fbadab17ad2)

2. Because of mess with items actions (some of them was defined in the list's actions file, some in other files), I made some cleanup. I created Items directory in list/components and moved there (via git mv) files and directories related to items. I created a model directory and put in it action.js, actionTypes.js reducer.js files.

![cloning_item](https://user-images.githubusercontent.com/25374390/56434627-aeb27d80-62d5-11e9-8769-c3f194dad8b4.gif)

adamklepacz commented on client/src/common/components/CardItem.scss:
```diff
@@ -31,6 +31,8 @@
       height: 12px;
       width: 12px;
+      min-height: 12px;
+      min-width: 12px;
```
is this necessary when height and width are 'hard set' above??

alemikolo replied:
Yes, it's necessary. Without these rules, if you add a long list title, the ellipsis is added but lock icon is very small (both in Chrome and Firefox):
![Firefox_Screenshot_2019-04-23T05-37-32 262Z](https://user-images.githubusercontent.com/25374390/56557264-b35b8800-659a-11e9-9a51-28f3344a17dc.png)

adamklepacz commented on client/src/common/components/Preloader.jsx:
```diff
@@ -12,34 +12,60 @@ export const PreloaderSize = {
+
+  render() {
+    const { message, size, theme } = this.props;
+
```
remove a line of the break here

adamklepacz commented on client/src/common/components/Toolbar/components/UserBar/UserBar.scss:
```diff
@@ -94,6 +94,7 @@
   }
 
   &__menu-logout {
+    font-family: inherit;
```
why do you need this?? 
Did font was different from other elements on the app??

alemikolo replied:
Without this rule, the logout button has a different font like the whole app. It's strange because it should inherit font-family by default, but it doesn't.

adamklepacz commented on client/src/modules/list/components/Items/ListItem/ListItem.jsx:
```diff
@@ -130,17 +147,45 @@ class ListItem extends PureComponent {
+          this.setState({ pending: false });
+          this.removePendingPromise(abortableCloning);
+        }
+      });
```
why this have to be done this way with 'abortable promises'
can it just be done like in any other actions ??

`cloneItem(listId, itemId).then(()=> //do something).catch(err)`

alemikolo replied:
If cloning the last item causes that the items count is above the limit of displayed items, this cloned item is being unmounted. Since there is still async action in progress we get warning in the browser. To prevent it I reject pending promises in componentWillUnmount. We want to do these actions only when the component is mounted so we have to check: if we created AbortPromiseException we know that component is unmounted and these actions shouldn't be invoked.

adamklepacz commented on client/src/modules/list/components/Items/ListItem/ListItem.scss:
```diff
@@ -153,6 +153,15 @@
+    justify-content: flex-start;
+    align-items: center;
+    min-height: 45px;
+    border-bottom: 1px solid #e6e7e8;
```
use variables

adamklepacz commented on client/src/modules/list/components/Items/model/reducer.js:
```diff
@@ -0,0 +1,58 @@
+          item._id === action.payload.item._id
+            ? {
+                ...action.payload.item,
+                voterIds: action.payload.item.voterIds
```
we need to check voting reducer cause I can see voterIds here. And we have removed IDS from front layer supposedly 😃

alemikolo replied:
You're right - it's unnecessary.  It looks like this is a souvenir from the past...

adamklepacz commented on server/controllers/list.js:
```diff
@@ -705,13 +705,62 @@ const updateItemDetails = (req, resp) => {
+    .then(list =>
+      resp.status(200).send({
+        message: 'Item successfully cloned.',
+        item: responseWithItem(list.items.id(newItemId))
```
I'm wondering do we need to send saved item, or we can just copy existing one from redux state and send new itemID only??

alemikolo replied:
We need new itemId and 'created-at', isVoted and VotesCount. Of course, I could send only these two first fields and set isVoted and VotesCount myself in cloneItem action but I think that now is more uniformly...

marekrozmus commented on client/src/common/components/CardItem.scss:
```diff
@@ -31,6 +31,8 @@
       height: 12px;
       width: 12px;
+      // min-height: 12px;
+      // min-width: 12px;
```
commentd out code added

marekrozmus commented on client/src/common/components/Preloader.jsx:
```diff
@@ -12,34 +12,59 @@ export const PreloaderSize = {
+    super(props);
+
+    this.state = { displayAnimation: false };
+    this.timer = setTimeout(this.enableAnimation, 250);
```
shouldn't this be rather in component did mount?

marekrozmus commented on client/src/modules/list/components/Items/ListItem/ListItem.jsx:
```diff
@@ -33,14 +37,27 @@ class ListItem extends PureComponent {
+    this.pendingPromises.map(promise => promise.abort());
+  }
+
+  appendPendingPromise = promise => {
```
why not `add` but `append`?

marekrozmus commented on client/src/modules/list/components/Items/model/actionTypes.js:
```diff
@@ -1,13 +1,16 @@
   ADD_SUCCESS: 'item/ADD_SUCCESS',
+  CLONE_FAILURE: 'item/CLONE__FAILURE',
+  CLONE_REQUEST: 'item/CLONE__REQUEST',
+  CLONE_SUCCESS: 'item/CLONE__SUCCESS',
```
why double `_`?

marekrozmus commented on client/src/modules/list/components/Items/model/actions.js:
```diff
@@ -0,0 +1,190 @@
+    authorName: updatedAuthor,
+    itemId,
+    isOrdered: !isOrdered,
+    listId
```
do you need to pass all that data to update `isOrdered`?
`itemId` and `isOrdered` should be enough - `listId` is already in api path

alemikolo replied:
I removed "listId". I removed updatedAuthor from passing it to backend (it's not used there now) but I left it in action. This "toggling item" functionality will be refactored in the future task "EOC-151 Fix item toggling (restoring should change author)" because for now it only changes "isOrdered" field. We should fix it to store new User Id when an item is restored, so "updatedAuthor" variable will be needed then.

alemikolo commented:
3. I added preloader animation to cloning item functionality. I added some delay to it by the way. In my first solution, I placed this delay inside preloader component but it had some UX's disadvantages - short flashes of content: the content was displayed next was disappearing and next, the preloader was displaying). If the network was fast, the preloader wasn't rendered but the content still flashed. 
In the current approach, I set "pending" prop after some time. If async action performed faster I clear timeout. Why 1 second: (https://ux.stackexchange.com/a/104782)

adamklepacz commented on client/src/modules/list/components/Items/ListItem/ListItem.jsx:
```diff
@@ -33,14 +37,27 @@ class ListItem extends PureComponent {
+    this.pendingPromises.map(promise => promise.abort());
+  }
+
+  addPendingPromise = promise => {
```
unecessary braces

adamklepacz commented on client/src/modules/list/components/Items/ListItem/ListItem.jsx:
```diff
@@ -130,17 +147,45 @@ class ListItem extends PureComponent {
+    } = this.props;
+
+    this.setState({ pending: true });
+    const abortableCloning = makeAbortablePromise(cloneItem(listId, itemId));
```
add line break before and after const

adamklepacz commented on client/src/modules/list/components/Items/model/actions.js:
```diff
@@ -0,0 +1,188 @@
+    itemId
+  })
+    .then(resp => resp.json())
+    .then(data => {
```
follow the same pattern as we did in previous functions 
so `json`   -->
`dispatch(cloneItemSuccess(listId, data.item));`

adamklepacz commented on server/controllers/list.js:
```diff
@@ -705,13 +705,62 @@ const updateItemDetails = (req, resp) => {
+      }
+
+      const { description, link, name } = list.items.id(itemId);
+
```
remove break line

adamklepacz commented on server/controllers/list.js:
```diff
@@ -705,13 +705,62 @@ const updateItemDetails = (req, resp) => {
+    )
+    .catch(err => {
+      if (err instanceof BadRequestException) {
+        const { status, message } = err;
```
add line break
